File: 1-TwoSum/1-TwoSum.java
// Last updated: 5/20/2026, 11:38:00 PM
class Solution {
    void putIntoHash(int[] nums, HashMap<Integer, Integer> hs){
        for (int i = 0; i < nums.length; i++) {
            hs.put(nums[i], i);
        }
    }
    public int[] twoSum(int[] nums, int target) {
        

        HashMap<Integer, Integer> hm = new HashMap<>(nums.length);
        putIntoHash(nums, hm);
        
        
        
        
        
        
        for (int i = 0; i < nums.length; i++) {
            if (hm.containsKey(target - nums[i]) && hm.get(target - nums[i]) != i){
                return new int[]{i,  hm.get(target - nums[i])};
            }
        }

        return new int[]{-1, -1};
    }


}

File: 1-TwoSum/1-TwoSum.dart
// Last updated: 5/20/2026, 11:37:58 PM
class Solution {
  List<int> twoSum(List<int> nums, int target) {
    // Create a list of pairs (value, original index)
    List<MapEntry<int, int>> indexedNums = [];
    for (int i = 0; i < nums.length; ++i) {
      indexedNums.add(MapEntry(nums[i], i));
    }

    // Sort the indexedNums by the values (nums[i])
    indexedNums.sort((a, b) => a.key.compareTo(b.key));

    int left = 0;
    int right = indexedNums.length - 1;

    while (left < right) {
      int sum = indexedNums[left].key + indexedNums[right].key;
      if (sum == target) {
        // Return the original indices
        return [indexedNums[left].value, indexedNums[right].value];
      } else if (sum < target) {
        left++;
      } else {
        right--;
      }
    }
    return []; // No solution found
  }
}
